### PR TITLE
DEV: rework the chat-live-pane

### PR DESCRIPTION
**Pull request by jjaffeux**
This PR is introducing glimmer usage in the chat-live-pane, for components but also for models. RestModel usage has been dropped in favor of native classes.

Other changes/additions in this PR:

- sticky dates, scrolling will now keep the date separator of the current section at the top of the screen
- better unread management, marking a channel  as unread will correctly mark the correct message and not mark the whole channel as read. Tracking state will also now correctly return unread count and unread mentions.
- adds an animation on bottom arrow
- better scrolling behavior, we should now always correctly keep the scroll position while loading more
- reactions are now more reactive, and will update their tooltip without needed to close/reopen it
- skeleton has been improved with placeholder images and reactions
- when making a reaction on the desktop message actions, the menu won't move anymore

### CODE REVIEW

File: plugins/chat/assets/javascripts/discourse/components/chat-live-pane.js [hunk 1]
```diff
@@ -458,62 +442,6 @@ export default Component.extend({
     );
   },
 
-  _prepareSingleMessage(messageData, previousMessageData) {
```
Comment (martin-brennan) [inline plugins/chat/assets/javascripts/discourse/components/chat-live-pane.js:461]: This is beautiful 🥲

File: plugins/chat/assets/javascripts/discourse/models/chat-message.js [hunk 2]
```diff
@@ -1,17 +1,46 @@
...
+  @tracked stagedId;
+  @tracked chatChannelId;
+  @tracked chatWebhookEvent;
+  @tracked cooked;
+  @tracked createdAt;
+  @tracked uploads;
+  @tracked excerpt;
+  @tracked message;
+  @tracked threadId;
+  @tracked reactions = new TrackedObject();
```
Comment (martin-brennan) [inline plugins/chat/assets/javascripts/discourse/models/chat-message.js:18]: Careful with this, since this is objects nested into objects I had to construct this myself to make it so all the tracked data worked correctly:

https://github.com/discourse/discourse/pull/20324/files#diff-97cdefa57def4cf645916acfa1c90a055cff3e567e0709879e8408bfc191ccc3R16-R23

**Review by martin-brennan — COMMENTED** [review]
Looking great so far, really looking forward to this :+1:

File: plugins/chat/app/models/chat_message.rb [hunk 3]
```diff
@@ -223,7 +223,7 @@ def full_url
   end
 
   def url
-    "/chat/message/#{self.id}"
+    "/chat/c/-/#{self.chat_channel_id}/#{self.id}"
```
Comment (jjaffeux) [inline plugins/chat/app/models/chat_message.rb:226]: It's not really part of this PR, but I noticed we were not using the new message URL here

File: plugins/chat/app/serializers/chat_channel_serializer.rb [hunk 4]
```diff
@@ -110,6 +110,7 @@ def current_user_membership
   def meta
     {
       message_bus_last_ids: {
+        channel_message_bus_last_id: MessageBus.last_id("/chat/#{object.id}"),
```
Comment (jjaffeux) [inline plugins/chat/app/serializers/chat_channel_serializer.rb:113]: This was causing some issues in tests as we were getting the id only once we had loaded messages

File: plugins/chat/app/services/chat_publisher.rb [hunk 5]
```diff
@@ -12,7 +12,7 @@ def self.publish_new!(chat_channel, chat_message, staged_id)
         { scope: anonymous_guardian, root: :chat_message },
       ).as_json
     content[:type] = :sent
-    content[:stagedId] = staged_id
+    content[:staged_id] = staged_id
```
Comment (jjaffeux) [inline plugins/chat/app/services/chat_publisher.rb:15]: generally trying to be very strict on casing difference between backend and frontend

File: plugins/chat/assets/javascripts/discourse/components/chat-draft-channel-screen.hbs [hunk 6]
```diff
@@ -19,9 +19,6 @@
   />
 
   {{#if this.previewedChannel}}
-    <ChatLivePane
-      @chatChannel={{this.previewedChannel}}
-      @includeHeader={{false}}
-    />
+    <ChatLivePane @channel={{this.previewedChannel}} @includeHeader={{false}} />
```
Comment (jjaffeux) [inline plugins/chat/assets/javascripts/discourse/components/chat-draft-channel-screen.hbs:22]: Generally trying to always prefer "channel" over "chatChannel"

File: plugins/chat/assets/javascripts/discourse/components/chat-live-pane.hbs [hunk 7]
```diff
@@ -1,148 +1,115 @@
...
+          @setReplyTo={{this.setReplyTo}}
+          @replyMessageClicked={{this.replyMessageClicked}}
+          @editButtonClicked={{this.editButtonClicked}}
+          @selectingMessages={{this.selectingMessages}}
+          @onStartSelectingMessages={{this.onStartSelectingMessages}}
+          @onSelectMessage={{this.onSelectMessage}}
+          @bulkSelectMessages={{this.bulkSelectMessages}}
+          @isHovered={{eq message.id this.hoveredMessageId}}
+          @onHoverMessage={{this.onHoverMessage}}
+          @resendStagedMessage={{this.resendStagedMessage}}
```
Comment (jjaffeux) [inline plugins/chat/assets/javascripts/discourse/components/chat-live-pane.hbs:65]: TODO: try this one

File: plugins/chat/assets/javascripts/discourse/components/chat-full-page-header.hbs [hunk 8]
```diff
@@ -0,0 +1,40 @@
+<div
```
Comment (martin-brennan) [inline plugins/chat/assets/javascripts/discourse/components/chat-full-page-header.hbs:1]: So glad you did this too, was thinking about moving this :pray:

File: plugins/chat/assets/javascripts/discourse/components/chat-live-pane.hbs (hunk 7 above)
Comment (martin-brennan) [reply plugins/chat/assets/javascripts/discourse/components/chat-live-pane.hbs:65]: Tried it locally using the Network tab to simulate no connection, worked fine.

File: plugins/chat/assets/javascripts/discourse/components/chat-live-pane.js [hunk 9]
```diff
@@ -203,163 +153,188 @@ export default Component.extend({
...
-  },
+  }
+
+  get capabilities() {
+    return getOwner(this).lookup("capabilities:main");
+  }
 
   @debounce(100)
-  fetchMessages(channel, options = {}) {
+  fetchMessages(options = {}) {
```
Comment (martin-brennan) [inline plugins/chat/assets/javascripts/discourse/components/chat-live-pane.js:170]: I guess the Thread component would have its own imlementation of `fetchMessages` and `_fetchMoreMessages` ? Then it would have a similar mechanic for clearMessages, appendMessages, etc on the `channel.activeThread` instead?

File: plugins/chat/assets/javascripts/discourse/models/chat-channel.js [hunk 10]
```diff
@@ -105,6 +115,53 @@ export default class ChatChannel extends RestModel {
...
+    }
+    if (details.can_load_more_past !== undefined) {
+      this.canLoadMorePast = details.can_load_more_past;
+    }
+    this.userSilenced = details.user_silenced ?? false;
+    this.status = details.channel_status ?? "open";
+    this.channelMessageBusLastId = details.channel_message_bus_last_id;
+  }
+
+  clearMessages() {
```
Comment (martin-brennan) [inline plugins/chat/assets/javascripts/discourse/models/chat-channel.js:134]: As I said earlier, we will likely also need all this kind of stuff on the Thread JS model, its good we have a framework for it now. Maybe this could be some mixin class we can extend so both can use it, since these 6 functions are pretty generic? Something like `HasChatMessages`

File: plugins/chat/assets/javascripts/discourse/models/chat-message.js [hunk 11]
```diff
@@ -1,26 +1,245 @@
...
+    return (
+      a.getFullYear() === b.getFullYear() &&
+      a.getMonth() === b.getMonth() &&
+      a.getDate() === b.getDate()
+    );
+  }
+}
+
+// https://stackoverflow.com/a/8809472
+function generateUDID() {
```
Comment (martin-brennan) [inline plugins/chat/assets/javascripts/discourse/models/chat-message.js:203]: We already have this in the codebase https://github.com/discourse/discourse/blob/61e2131738fde0dc72dd03d5b091d94dbf559e72/app/assets/javascripts/pretty-text/addon/guid.js . Since these pretty-text helpers can import from discourse-common/lib, we should probably make a new file for this to export the `generateUUID()` function, similar to https://github.com/discourse/discourse/blob/61e2131738fde0dc72dd03d5b091d94dbf559e72/app/assets/javascripts/discourse-common/addon/lib/dom-from-string.js

File: plugins/chat/assets/javascripts/discourse/modifiers/chat/track-highlighted.js [hunk 12]
```diff
@@ -0,0 +1,59 @@
+import Modifier from "ember-modifier";
+import { registerDestructor } from "@ember/destroyable";
+import { bind } from "discourse-common/utils/decorators";
+
+export default class ChatTrackHighlighted extends Modifier {
```
Comment (martin-brennan) [inline plugins/chat/assets/javascripts/discourse/modifiers/chat/track-highlighted.js:5]: This is cool!

File: plugins/chat/assets/javascripts/discourse/modifiers/chat/track-message.js [hunk 13]
```diff
@@ -0,0 +1,39 @@
+import Modifier from "ember-modifier";
+import { registerDestructor } from "@ember/destroyable";
+import { bind } from "discourse-common/utils/decorators";
+
+export default class ChatTrackMessage extends Modifier {
```
Comment (martin-brennan) [inline plugins/chat/assets/javascripts/discourse/modifiers/chat/track-message.js:5]: Why is this needed when there is already `track-message-visibility.js`, they both do the same thing from what I can see? I am probably wrong, just curious

File: plugins/chat/assets/javascripts/discourse/services/chat-api.js [hunk 14]
```diff
@@ -233,6 +233,39 @@ export default class ChatApi extends Service {
     );
   }
 
+  /**
+   * Returns messages of a channel, from the last message or a specificed target.
+   * @param {number} channelId - The ID of the channel.
+   * @param {object} data - Params of the query.
+   * @param {integer} data.targetMessageId - ID of the targeted message.
```
Comment (martin-brennan) [inline plugins/chat/assets/javascripts/discourse/services/chat-api.js:240]: I guess we could add `threadId` to this list later on too

**Comment by martin-brennan** [comment]
Working quite well, the main issue I had is that _sometimes_ (sorry it's quite hard to reproduce) the message I just sent sticks around in the composer for ~1s or a bit more, even though the new message has been put in the channel, its kind of jarring. It could just be me, not sure. Left some questions too.

Everything else works fine, I tried all the different message actions etc. As you said only drafts doesn't work.

Love all the things that you have moved into their own components and made proper classes with more `@tracked` , will make things much easier. And the skeleton and loading indicator changes are :ok_hand:

**Comment by martin-brennan** [comment]
> Working quite well, the main issue I had is that sometimes (sorry it's quite hard to reproduce) the message I just sent sticks around in the composer for ~1s or a bit more, even though the new message has been put in the channel, its kind of jarring. It could just be me, not sure.

@jjaffeux ignore this, it's some other issue in core causing things to go slow locally.

File: plugins/chat/assets/javascripts/discourse/models/chat-message.js (hunk 11 above)
Comment (jjaffeux) [reply plugins/chat/assets/javascripts/discourse/models/chat-message.js:203]: Hmm nice didn't know, will look to use it :+1:

File: plugins/chat/app/queries/chat_channel_unreads_query.rb [hunk 15]
```diff
@@ -0,0 +1,39 @@
...
+      AND user_chat_channel_memberships.user_id = :user_id
+      WHERE NOT read
+      AND notifications.user_id = :user_id
+      AND notifications.notification_type = :notification_type
+      AND (data::json->>'chat_message_id')::bigint > COALESCE(user_chat_channel_memberships.last_read_message_id, 0)
+      AND (data::json->>'chat_channel_id')::bigint = :channel_id
+    ) AS mention_count;
+    SQL
+
+    builder
```
Comment (martin-brennan) [inline plugins/chat/app/queries/chat_channel_unreads_query.rb:30]: I'm not sure all this is necessary...couldn't you just do this?

```
sql =
  <<~SQL,
      SELECT (
      SELECT COUNT(*) AS unread_count
      FROM chat_messages
      INNER JOIN chat_channels ON chat_channels.id = chat_messages.chat_channel_id
      INNER JOIN user_chat_channel_memberships ON user_chat_channel_memberships.chat_channel_id = chat_channels.id
      WHERE chat_channels.id = :channel_id
      AND chat_messages.user_id != :user_id
      AND user_chat_channel_memberships.user_id = :user_id
      AND chat_messages.id > COALESCE(user_chat_channel_memberships.last_read_message_id, 0)
      AND chat_messages.deleted_at IS NULL
    ) AS unread_count,
    (
      SELECT COUNT(*) AS mention_count
      FROM notifications
      INNER JOIN user_chat_channel_memberships ON user_chat_channel_memberships.chat_channel_id = :channel_id
      AND user_chat_channel_memberships.user_id = :user_id
      WHERE NOT read
      AND notifications.user_id = :user_id
      AND notifications.notification_type = :notification_type
      AND (data::json->>'chat_message_id')::bigint > COALESCE(user_chat_channel_memberships.last_read_message_id, 0)
      AND (data::json->>'chat_channel_id')::bigint = :channel_id
    ) AS mention_count;
    SQL
  DB
    .query(
      sql,
      channel_id: channel_id,
      user_id: user_id,
      notification_type: Notification.types[:chat_mention],
    )
    .first
    .to_h
```

This will always return `{:unread_count=>0, :mention_count=>0}` (with the correct counts, and 0 if there really are none)

Comment (martin-brennan) [reply plugins/chat/app/queries/chat_channel_unreads_query.rb:30]: There is no need for `query_hash` , safe nav, or `symbolize_keys` then

Comment (jjaffeux) [reply plugins/chat/app/queries/chat_channel_unreads_query.rb:30]: Yes sure cool :+1: I tried something like this but didn't get what I wanted, must have done an error. THanks.

File: plugins/chat/assets/javascripts/discourse/initializers/chat-cook-function.js [hunk 16]
```diff
@@ -0,0 +1,32 @@
...
+        site.markdown_additional_options?.chat?.limited_pretty_text_features,
+      markdownItRules:
+        site.markdown_additional_options?.chat
+          ?.limited_pretty_text_markdown_rules,
+      hashtagTypesInPriorityOrder: site.hashtag_configurations["chat-composer"],
+      hashtagIcons: site.hashtag_icons,
+    };
+
+    generateCookFunction(markdownOptions).then((cookFunction) => {
+      ChatMessage.cookFunction = (raw) => {
```
Comment (martin-brennan) [inline plugins/chat/assets/javascripts/discourse/initializers/chat-cook-function.js:24]: Excellent, I can't remember if I pointed this out or only saw it, but this avoids so much unnecessary work, nice!

File: plugins/chat/assets/javascripts/discourse/models/chat-channel.js [hunk 17]
```diff
@@ -105,6 +116,57 @@ export default class ChatChannel extends RestModel {
...
+    this.canFlag = details.can_flag ?? false;
+    this.canModerate = details.can_moderate ?? false;
+    if (details.can_load_more_future !== undefined) {
+      this.canLoadMoreFuture = details.can_load_more_future;
+    }
+    if (details.can_load_more_past !== undefined) {
+      this.canLoadMorePast = details.can_load_more_past;
+    }
+    this.userSilenced = details.user_silenced ?? false;
+    this.status = details.channel_status ?? "open";
```
Comment (martin-brennan) [inline plugins/chat/assets/javascripts/discourse/models/chat-channel.js:135]: This should always have a value, no?

File: plugins/chat/assets/javascripts/discourse/models/chat-channel.js [hunk 18]
```diff
@@ -142,12 +208,17 @@ ChatChannel.reopenClass({
...
-    args.chatableType = args.chatable_type;
-    args.membershipsCount = args.memberships_count;
+    this._remapKey(args, "chatable_type", "chatableType");
+    this._remapKey(args, "memberships_count", "membershipsCount");
+    this._remapKey(args, "last_message_sent_at", "lastMessageSentAt");
 
     return this._super(args);
   },
 
+  _remapKey(obj, oldKey, newKey) {
```
Comment (martin-brennan) [inline plugins/chat/assets/javascripts/discourse/models/chat-channel.js:218]: This could be useful to extract for other uses. Also, why not use this for `set details(details)` above?

File: plugins/chat/assets/javascripts/discourse/models/chat-message-draft.js [hunk 19]
```diff
@@ -0,0 +1,62 @@
+import { tracked } from "@glimmer/tracking";
+
+export default class ChatMessageDraft {
+  static create(args = {}) {
```
Comment (martin-brennan) [inline plugins/chat/assets/javascripts/discourse/models/chat-message-draft.js:4]: Why bother with this when we can just do `new ChatMessageDraft`?

File: plugins/chat/assets/javascripts/discourse/models/chat-message-draft.js [hunk 20]
```diff
@@ -0,0 +1,62 @@
...
+    return new ChatMessageDraft(args ?? {});
+  }
+
+  @tracked uploads;
+  @tracked message;
+  @tracked _replyToMsg;
+
+  constructor(args = {}) {
```
Comment (martin-brennan) [inline plugins/chat/assets/javascripts/discourse/models/chat-message-draft.js:12]: Not a huge deal, but this could be written as 

```javascript
  constructor({ message = "", uploads = [], replyToMsg } = {}) {
    this.message = message;
    this.uploads = uploads;
    this.replyToMsg = replyToMsg;
  }
```

I just remembered this from David's post about the destructuring

File: plugins/chat/assets/javascripts/discourse/models/chat-message-reaction.js [hunk 21]
```diff
@@ -0,0 +1,33 @@
+import { tracked } from "@glimmer/tracking";
+import User from "discourse/models/user";
+import { TrackedArray } from "@ember-compat/tracked-built-ins";
+
+export default class ChatMessageReaction {
+  static create(args = {}) {
```
Comment (martin-brennan) [inline plugins/chat/assets/javascripts/discourse/models/chat-message-reaction.js:6]: Same wondering about why this is needed when we can just do `new ChatMessageReaction`

File: plugins/chat/assets/javascripts/discourse/models/chat-message-reaction.js [hunk 22]
```diff
@@ -0,0 +1,33 @@
...
+    return new ChatMessageReaction(args);
+  }
+
+  @tracked count = 0;
+  @tracked reacted = false;
+  @tracked users = [];
+
+  constructor(args = {}) {
```
Comment (martin-brennan) [inline plugins/chat/assets/javascripts/discourse/models/chat-message-reaction.js:14]: Same destructuring could be done here (with the exception of `initUserModels`)

File: plugins/chat/assets/javascripts/discourse/models/chat-message.js [hunk 23]
```diff
@@ -1,26 +1,193 @@
...
-  },
+  static createStagedMessage(channel, args = {}) {
+    args.staged_id = guid();
+    return new ChatMessage(channel, args);
+  }
 
-  _initReactions(args) {
-    args.reactions = EmberObject.create(args.reactions || {});
-  },
+  @tracked id;
```
Comment (martin-brennan) [inline plugins/chat/assets/javascripts/discourse/models/chat-message.js:21]: They really need a `@trackAllOfThisStuff(a, b, c)` API :laughing:

File: plugins/chat/assets/javascripts/discourse/models/chat-message.js [hunk 24]
```diff
@@ -1,26 +1,193 @@
...
+  #calendarDate(date) {
+    return moment(date).calendar(moment(), {
+      sameDay: `[${I18n.t("chat.chat_message_separator.today")}]`,
+      lastDay: `[${I18n.t("chat.chat_message_separator.yesterday")}]`,
+      lastWeek: "LL",
+      sameElse: "LL",
+    });
+  }
+
+  @cached
```
Comment (martin-brennan) [inline plugins/chat/assets/javascripts/discourse/models/chat-message.js:107]: I did not realise that these `get`ters did not cache by default, I thought they were like computed properties and only changed when something it uses changes, e.g. `this.channel.messages`. Good to know.

**Review by martin-brennan — APPROVED** [review]
Looks great, mostly I have minor comments/nits that can be changed later or rejected by you anyway. Ship it!

**Comment by martin-brennan** [comment]
Oh also only 2 failing system specs, the failing backend one is some unrelated theme spec

File: plugins/chat/assets/javascripts/discourse/models/chat-message.js (hunk 23 above)
Comment (jjaffeux) [reply plugins/chat/assets/javascripts/discourse/models/chat-message.js:21]: could be handy indeed

File: plugins/chat/assets/javascripts/discourse/components/chat-message.hbs [hunk 25]
```diff
@@ -40,19 +41,22 @@
   {{did-insert this.setMessageActionsAnchors}}
   {{did-insert this.decorateCookedMessage}}
   {{did-update this.decorateCookedMessage @message.id}}
+  {{did-update this.decorateCookedMessage @message.version}}
```
[the remark this answers is not in the document]
Comment (jjaffeux) [reply plugins/chat/assets/javascripts/discourse/components/chat-message.hbs:44]: I just think it's cleaner but yes it's a similar pattern.

File: plugins/chat/assets/javascripts/discourse/models/chat-channel.js (hunk 17 above)
Comment (jjaffeux) [reply plugins/chat/assets/javascripts/discourse/models/chat-channel.js:135]: Yes you must be right, will just remove it :+1:

File: plugins/chat/assets/javascripts/discourse/models/chat-channel.js (hunk 18 above)
Comment (jjaffeux) [reply plugins/chat/assets/javascripts/discourse/models/chat-channel.js:218]: It will be all gone when I move chat-channel model to the same pattern used in chat-message

File: plugins/chat/assets/javascripts/discourse/models/chat-message-draft.js (hunk 19 above)
Comment (jjaffeux) [reply plugins/chat/assets/javascripts/discourse/models/chat-message-draft.js:4]: It's mostly for history, we have always created our models this way and I think it's nice, it has some similarities with rails too. You can easily created dedicated creator if needed too.

File: plugins/chat/assets/javascripts/discourse/models/chat-message-reaction.js (hunk 22 above)
Comment (jjaffeux) [reply plugins/chat/assets/javascripts/discourse/models/chat-message-reaction.js:14]: Yes same reason than above, not super fan of destructuring in things which can be long lists

File: plugins/chat/assets/javascripts/discourse/models/chat-message-reaction.js (hunk 21 above)
Comment (jjaffeux) [reply plugins/chat/assets/javascripts/discourse/models/chat-message-reaction.js:6]: see https://github.com/discourse/discourse/pull/20397#discussion_r1122750585

File: plugins/chat/assets/javascripts/discourse/components/channels-list.js [hunk 26]
```diff
@@ -1,24 +1,30 @@
...
+  get showMobileDirectMessageButton() {
+    return this.site.mobileView && this.showDirectMessageChannels;
+  }
+
+  get inSidebar() {
+    return this.args.inSidebar ?? false;
+  }
+
+  get publicMessageChannelsEmpty() {
+    return this.chatChannelsManager.publicMessageChannels?.length === 0;
```
[the remark this answers is not in the document]
Comment (jjaffeux) [reply plugins/chat/assets/javascripts/discourse/components/channels-list.js:25]: The whole hierarchy is though to always be an empty array. It starts from something like this:

```
Object.values({})
```

File: plugins/chat/app/services/chat_publisher.rb [hunk 27]
```diff
@@ -133,9 +133,13 @@ def self.user_tracking_state_message_bus_channel(user_id)
   end
 
   def self.publish_user_tracking_state(user, chat_channel_id, chat_message_id)
+    data = { chat_channel_id: chat_channel_id.to_i, chat_message_id: chat_message_id.to_i }.merge(
```
[the remark this answers is not in the document]
Comment (jjaffeux) [reply plugins/chat/app/services/chat_publisher.rb:136]: You are probably right :+1:

File: plugins/chat/app/serializers/chat_message_serializer.rb [hunk 28]
```diff
@@ -35,23 +35,23 @@ def excerpt
...
       .reactions
       .group_by(&:emoji)
-      .each do |emoji, reactions|
-        users = reactions[0..5].map(&:user).filter { |user| user.id != scope&.user&.id }[0..4]
-
+      .map do |emoji, reactions|
         next unless Emoji.exists?(emoji)
 
-        reactions_hash[emoji] = {
+        users = reactions[0..5].map(&:user)[0..5]
```
[the remark this answers is not in the document]
Comment (jjaffeux) [reply plugins/chat/app/serializers/chat_message_serializer.rb:44]: :+1:

File: plugins/chat/app/queries/chat_channel_unreads_query.rb [hunk 29]
```diff
@@ -0,0 +1,40 @@
+# frozen_string_literal: true
+
+class ChatChannelUnreadsQuery
+  def self.call(channel_id, user_id)
```
[the remark this answers is not in the document]
Comment (jjaffeux) [reply plugins/chat/app/queries/chat_channel_unreads_query.rb:4]: Yes right, will do the change :+1:

File: plugins/chat/app/serializers/chat_message_serializer.rb [hunk 30]
```diff
@@ -35,23 +35,23 @@ def excerpt
...
       .reactions
       .group_by(&:emoji)
-      .each do |emoji, reactions|
-        users = reactions[0..5].map(&:user).filter { |user| user.id != scope&.user&.id }[0..4]
-
+      .map do |emoji, reactions|
         next unless Emoji.exists?(emoji)
 
-        reactions_hash[emoji] = {
+        users = reactions.take(5).map(&:user).take(5)
```
Comment (CvX) [inline plugins/chat/app/serializers/chat_message_serializer.rb:44]: The second `take()` doesn't do anything, does it? It's not possible for `map()` to increase the size of an array?

Comment (jjaffeux) [reply plugins/chat/app/serializers/chat_message_serializer.rb:44]: Yes I think you are right, the original code was already convulted :+1:
